components/scheduler2/scheduler.go

```diff
@@ -0,0 +1,226 @@
+		},
+		cacheConflictToken: scheduler.InitialConflictToken,
+		compiledSpec:       nil,
+	}
+}
+
+// RegisterStateMachine registers state machine definitions with the HSM
+// registry. Should be called during dependency injection.
+func RegisterStateMachines(r *hsm.Registry) error {
+	if err := r.RegisterMachine(schedulerMachineDefinition{}); err != nil {
+		return err
+	}
+	if err := r.RegisterMachine(generatorMachineDefinition{}); err != nil {
+		return err
+	}
+	if err := r.RegisterMachine(executorMachineDefinition{}); err != nil {
+		return err
+	}
+	// TODO: add other state machines here
+	return nil
+}
+
+// MachineCollection creates a new typed [statemachines.Collection] for operations.
+func MachineCollection(tree *hsm.Node) hsm.Collection[Scheduler] {
+	return hsm.NewCollection[Scheduler](tree, SchedulerMachineType)
+}
+
+func (s Scheduler) State() enumsspb.Scheduler2State {
+	return s.SchedulerInternal.State
+}
+
+func (s Scheduler) SetState(state enumsspb.Scheduler2State) {
+	s.SchedulerInternal.State = state
+}
+
+func (s Scheduler) RegenerateTasks(node *hsm.Node) ([]hsm.Task, error) {
+	// The top level scheduler has no tasks of its own.
+	return nil, nil
+}
+
+func (schedulerMachineDefinition) Type() string {
+	return SchedulerMachineType
+}
+
+func (schedulerMachineDefinition) Serialize(state any) ([]byte, error) {
+	if state, ok := state.(Scheduler); ok {
+		return proto.Marshal(state.SchedulerInternal)
+	}
+	return nil, fmt.Errorf("invalid scheduler state provided: %v", state)
+}
+
+func (schedulerMachineDefinition) Deserialize(body []byte) (any, error) {
+	state := &schedspb.SchedulerInternal{}
+	return Scheduler{
+		SchedulerInternal: state,
+		compiledSpec:      nil,
+	}, proto.Unmarshal(body, state)
+}
+
+func (schedulerMachineDefinition) CompareState(a any, b any) (int, error) {
+	panic("TODO: CompareState not yet implemented for Scheduler")
+}
+
+// UseScheduledAction returns true when the Scheduler should allow scheduled
+// actions to be taken.
+//
+// When decrement is true, the schedule's state's `RemainingActions` counter is
+// decremented when an action can be taken. When decrement is false, no state
+// is mutated.
+func (s Scheduler) UseScheduledAction(decrement bool) bool {
+	// If paused, don't do anything.
+	if s.Schedule.State.Paused {
+		return false
+	}
+
+	// If unlimited actions, allow.
+	if !s.Schedule.State.LimitedActions {
+		return true
+	}
+
+	// Otherwise check and decrement limit.
+	if s.Schedule.State.RemainingActions > 0 {
+		if decrement {
+			s.Schedule.State.RemainingActions--
+
+			// The conflict token is updated because a client might be in the process of
+			// preparing an update request that increments their schedule's RemainingActions
+			// field.
+			s.UpdateConflictToken()
+		}
+		return true
+	}
+
+	// No actions left
+	return false
+}
+
+func (s Scheduler) CompiledSpec(specBuilder *scheduler.SpecBuilder) (*scheduler.CompiledSpec, error) {
```

Comment:
all these receivers have to be pointers, no? specifically you can't usefully store to s.compiledSpec otherwise
```suggestion
func (s *Scheduler) CompiledSpec(specBuilder *scheduler.SpecBuilder) (*scheduler.CompiledSpec, error) {
```